Fix race condition when calculating the LSO

mmaslankaprv opened the pull request:
<!--
See https://github.com/redpanda-data/redpanda/blob/dev/CONTRIBUTING.md#pull-request-body
for more details and examples of what is expected in a PR body.

Content in this top section is REQUIRED. Describe, in plain language, the motivation
behind the change (bug fix, feature, improvement) in this PR and how the included
commits address it.

Add the GitHub keyword `Fixes` to link to bug(s) this PR will fix, e.g.
  Fixes #ISSUE-NUMBER, Fixes #ISSUE-NUMBER, ...

If this PR is a backport, link to the original with `Backport of PR`, e.g.
  Backport of PR #PR-NUMBER
-->

## Backports Required

<!-- Checking at least one of the checkboxes is REQUIRED if this PR is not a backport. -->

- [ ] none - not a bug fix
- [ ] none - this is a backport
- [x] none - issue does not exist in previous branches
- [ ] none - papercut/not impactful enough to backport
- [ ] v25.3.x
- [ ] v25.2.x
- [ ] v25.1.x
- [ ] v24.3.x

## Release Notes

<!--
If the changes in this PR do not need to be mentioned in the release
notes, then don't add a sub-section and simply list `none`, e.g.

* none

Otherwise, adding a sub-section or `none` is REQUIRED if the PR is not a backport PR.
If this is a backport PR, adding contents to this section will override
the release notes section inherited from the original PR to dev.

Add one or more of the sub-sections with a short description bullet
point of the change, e.g.

### Bug Fixes

* Short description of the bug fix if this is a PR to `dev` branch.

### Features

* Short description of the feature. Explain how to configure.

### Improvements

* Short description of how this PR improves existing behavior.

-->
- none

bharathv committed:
tx/rm_stm: do not compute lso at shutdown

As the stm is shutting down, the state the function is looking at may
be partial and there is a chance LSO is overestimated which can include
open transactions.

bharathv committed:
tx/rm_stm: misc comment edits

bharathv committed:
tx/rm_stm: do not compute LSO when reset is in progress

When a reset is in progress (eg: raft snapshot), the stm may clear all
the inflight transactions and reset the next apply offset. Due to
scheduling points, if an LSO request comes in racily, it may not see a
consistent state of the stm.

This commit returns the last known lso as a conservative approach until
the reset is finished.

bharathv committed:
tx/rm_stm/tests: add a test for monotonicity of lso

bharathv committed:
tx/rm_stm: avoid a dangling rvalue reference from function arg

Move it on to the stack. A deeper fix is to pass it by copy but thats a
bigger change and can be done later.

bharathv committed:
tx/rm_stm: avoid overlfow in timeout.

Using long_max milliseconds will result in overflow during internal
duration casts. instead use a large-ish timeout equivalent to no
timeout.

bharathv committed:
tx/rm_stm: disable test_lso_bound_by_open_tx

mmaslankaprv commented:
/ci-repeat 1

vbotbuildovich commented:
<!--FAILURES SLASH COMMAND https://buildkite.com/redpanda/redpanda/builds/75633-->
<!--35d16873-ec28-423e-bc39-82a5f640a863-->
<!--9188b6ab-da6e-4aa5-8bb8-e179e5a47208-->
<!--1240d9c6-7125-4193-b9ac-35271af3fff0-->
<!--97da7c49-7f00-4977-9786-3f751e8e2084-->
<!--f338c4e8-674b-4b63-ab27-12da08d35b1e-->
<!--0c91d569-4623-4775-a6e5-1f1b5f36c87a-->
<!--de5fdbeb-9ffa-40f4-81ac-b760d527d707-->
<!--b6bf6ad0-a77e-41b7-a219-fe0635081770-->
<!--5eb74028-32fa-40f6-a6a5-751816b4bd96-->
<!--910f673e-d9c4-4cf1-8c9b-8e00a055dd13-->
<!--d06d79d7-e26c-4722-9ce8-0f97c9331e97-->
<!--9b7b20c5-2aa2-4865-81fa-01677cd7febe-->
<!--7f440229-aac0-4706-93b9-12a59ff0206c-->
<!--2b3d07b8-aaae-4b0f-bccd-072932936d9b-->
<!--1d4700d8-50ef-4f7e-be98-1026a6620aa7-->
<!--title-->
## Retry command for [Build#75633](https://buildkite.com/redpanda/redpanda/builds/75633)
please wait until all jobs are finished before running the slash command
<!--/title-->
<!--body-->
```
/ci-repeat 1
tests/rptest/tests/datalake/delayed_translation_test.py::DatalakeDelayedTranslationTest.test_basic@{"catalog_type":"nessie","cloud_storage_type":1,"query_engine":"trino"}
tests/rptest/tests/datalake/delayed_translation_test.py::DatalakeDelayedTranslationTest.test_basic@{"catalog_type":"rest_hadoop","cloud_storage_type":1,"query_engine":"trino"}
tests/rptest/tests/datalake/cluster_restore_test.py::DatalakeClusterRestoreTest.test_restore_partition_spec@{"catalog_type":"nessie","cloud_storage_type":1}
tests/rptest/tests/datalake/cluster_restore_test.py::DatalakeClusterRestoreTest.test_restore_partition_spec@{"catalog_type":"rest_hadoop","cloud_storage_type":1}
tests/rptest/tests/datalake/delayed_translation_test.py::DatalakeDelayedTranslationTest.test_basic@{"catalog_type":"rest_jdbc","cloud_storage_type":1,"query_engine":"trino"}
tests/rptest/tests/datalake/delayed_translation_test.py::DatalakeDelayedTranslationTest.test_basic@{"catalog_type":"nessie","cloud_storage_type":1,"query_engine":"spark"}
tests/rptest/tests/datalake/cluster_restore_test.py::DatalakeClusterRestoreTest.test_restore_partition_spec@{"catalog_type":"rest_jdbc","cloud_storage_type":1}
tests/rptest/tests/datalake/delayed_translation_test.py::DatalakeDelayedTranslationTest.test_basic@{"catalog_type":"rest_jdbc","cloud_storage_type":1,"query_engine":"spark"}
tests/rptest/tests/datalake/batching_test.py::DatalakeBatchingTest.test_batching@{"catalog_type":"rest_jdbc","cloud_storage_type":1,"expect_large_files":true,"query_engine":"spark"}
tests/rptest/tests/partition_force_reconfiguration_test.py::NodeWiseRecoveryTest.test_node_wise_recovery@{"dead_node_count":1}
tests/rptest/tests/datalake/cluster_restore_test.py::DatalakeClusterRestoreTest.test_basic@{"catalog_type":"rest_hadoop","cloud_storage_type":1}
tests/rptest/tests/e2e_iam_role_test.py::STSRoleFetchTests.test_write
tests/rptest/tests/datalake/delayed_translation_test.py::DatalakeDelayedTranslationTest.test_basic@{"catalog_type":"rest_hadoop","cloud_storage_type":1,"query_engine":"spark"}
tests/rptest/tests/shadow_indexing_compacted_topic_test.py::ShadowIndexingCompactedTopicTest.test_upload@{"cloud_storage_type":1}
tests/rptest/tests/e2e_shadow_indexing_test.py::EndToEndShadowIndexingTest.test_write@{"cloud_storage_type":2}
tests/rptest/tests/random_node_operations_smoke_test.py::RedpandaNodeOperationsSmokeTest.test_node_ops_smoke_test@{"cloud_storage_type":1,"mixed_versions":true}
tests/rptest/tests/shadow_indexing_compacted_topic_test.py::ShadowIndexingCompactedTopicTest.test_upload@{"cloud_storage_type":2}
tests/rptest/tests/random_node_operations_smoke_test.py::RedpandaNodeOperationsSmokeTest.test_node_ops_smoke_test@{"cloud_storage_type":1,"mixed_versions":false}
```
<!--/body-->

vbotbuildovich commented:
<!--TEST RESULTS-->
<!--28488e3a-6066-4033-8616-f2370b4463d9-->
<!--TEST RESULTS-->
<!--0cd6af7c-5f47-4ce5-b951-1d7129db2a0f-->
<!--TEST RESULTS-->
<!--9a7793cd-68fe-46fe-84cd-361066e68f1f-->
<!--TEST RESULTS-->
<!--58bcb083-f9bf-43c2-95b9-2be463a902f4-->
<!--TEST RESULTS-->
<!--bdb3b9b6-7f36-475b-a565-1b8bd0e62228-->
<!--title-->
## CI test results
<!--/title-->
<!--body-->
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/75633">build#75633</a></summary>


| test_class                                | test_method                                | test_arguments                                                                                              | test_kind   | job_url                                                                                   | test_status   | passed   | reason                                                                                                                                                                  | test_history                                                                                                                                                                          |
|:------------------------------------------|:-------------------------------------------|:------------------------------------------------------------------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|:------------------------------------------------------------------------------------------------------------------------------------------------------------------------|:--------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| DatalakeBatchingTest                      | test_batching                              | {"catalog_type": "rest_jdbc", "cloud_storage_type": 1, "expect_large_files": true, "query_engine": "spark"} | integration | https://buildkite.com/redpanda/redpanda/builds/75633#019a5380-ea75-43d0-8382-4c4fb8b6caeb | FLAKY         | 2/21     | upstream reliability is '100.0'. current run reliability is '9.523809523809524'. drift is 90.47619 and the allowed drift is set to 50. The test should FAIL             | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=DatalakeBatchingTest&test_method=test_batching                                      |
| DatalakeClusterRestoreTest                | test_basic                                 | {"catalog_type": "nessie", "cloud_storage_type": 1}                                                         | integration | https://buildkite.com/redpanda/redpanda/builds/75633#019a537c-398c-4fc0-a8df-a573f6d6447e | FLAKY         | 4/21     |                                                                                                                                                                         | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=DatalakeClusterRestoreTest&test_method=test_basic                                   |
| DatalakeClusterRestoreTest                | test_basic                                 | {"catalog_type": "rest_hadoop", "cloud_storage_type": 1}                                                    | integration | https://buildkite.com/redpanda/redpanda/builds/75633#019a537c-398d-4381-b5a7-b38dbbeb198a | FLAKY         | 5/21     | upstream reliability is '100.0'. current run reliability is '23.809523809523807'. drift is 76.19048 and the allowed drift is set to 50. The test should FAIL            | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=DatalakeClusterRestoreTest&test_method=test_basic                                   |
| DatalakeClusterRestoreTest                | test_basic                                 | {"catalog_type": "rest_jdbc", "cloud_storage_type": 1}                                                      | integration | https://buildkite.com/redpanda/redpanda/builds/75633#019a537c-398e-44b5-84a7-b60f2d295c2d | FLAKY         | 11/21    | upstream reliability is '100.0'. current run reliability is '52.38095238095239'. drift is 47.61905 and the allowed drift is set to 50. The test should PASS             | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=DatalakeClusterRestoreTest&test_method=test_basic                                   |
| DatalakeClusterRestoreTest                | test_basic                                 | {"catalog_type": "rest_jdbc", "cloud_storage_type": 1}                                                      | integration | https://buildkite.com/redpanda/redpanda/builds/75633#019a5380-ea77-4f96-8901-87be56c64e3b | FLAKY         | 20/21    | upstream reliability is '100.0'. current run reliability is '95.23809523809523'. drift is 4.7619 and the allowed drift is set to 50. The test should PASS               | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=DatalakeClusterRestoreTest&test_method=test_basic                                   |
| DatalakeClusterRestoreTest                | test_restore_partition_spec                | {"catalog_type": "nessie", "cloud_storage_type": 1}                                                         | integration | https://buildkite.com/redpanda/redpanda/builds/75633#019a537c-3984-4dfc-8417-c128e7e1fd05 | FLAKY         | 4/21     | upstream reliability is '100.0'. current run reliability is '19.047619047619047'. drift is 80.95238 and the allowed drift is set to 50. The test should FAIL            | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=DatalakeClusterRestoreTest&test_method=test_restore_partition_spec                  |
| DatalakeClusterRestoreTest                | test_restore_partition_spec                | {"catalog_type": "rest_hadoop", "cloud_storage_type": 1}                                                    | integration | https://buildkite.com/redpanda/redpanda/builds/75633#019a537c-3986-4c78-bd52-008b47d8f269 | FLAKY         | 8/21     | upstream reliability is '100.0'. current run reliability is '38.095238095238095'. drift is 61.90476 and the allowed drift is set to 50. The test should FAIL            | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=DatalakeClusterRestoreTest&test_method=test_restore_partition_spec                  |
| DatalakeClusterRestoreTest                | test_restore_partition_spec                | {"catalog_type": "rest_jdbc", "cloud_storage_type": 1}                                                      | integration | https://buildkite.com/redpanda/redpanda/builds/75633#019a537c-3987-441c-9366-014909806170 | FLAKY         | 4/21     | upstream reliability is '100.0'. current run reliability is '19.047619047619047'. drift is 80.95238 and the allowed drift is set to 50. The test should FAIL            | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=DatalakeClusterRestoreTest&test_method=test_restore_partition_spec                  |
| DatalakeDelayedTranslationTest            | test_basic                                 | {"catalog_type": "nessie", "cloud_storage_type": 1, "query_engine": "spark"}                                | integration | https://buildkite.com/redpanda/redpanda/builds/75633#019a537c-398c-4fc0-a8df-a573f6d6447e | FAIL          | 0/1      |                                                                                                                                                                         | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=DatalakeDelayedTranslationTest&test_method=test_basic                               |
| DatalakeDelayedTranslationTest            | test_basic                                 | {"catalog_type": "nessie", "cloud_storage_type": 1, "query_engine": "spark"}                                | integration | https://buildkite.com/redpanda/redpanda/builds/75633#019a5380-ea72-4f2d-babb-04e498f6fb45 | FLAKY         | 7/21     | upstream reliability is '100.0'. current run reliability is '33.33333333333333'. drift is 66.66667 and the allowed drift is set to 50. The test should FAIL             | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=DatalakeDelayedTranslationTest&test_method=test_basic                               |
| DatalakeDelayedTranslationTest            | test_basic                                 | {"catalog_type": "rest_hadoop", "cloud_storage_type": 1, "query_engine": "spark"}                           | integration | https://buildkite.com/redpanda/redpanda/builds/75633#019a537c-398d-4381-b5a7-b38dbbeb198a | FAIL          | 0/21     | The test has failed across all retries                                                                                                                                  | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=DatalakeDelayedTranslationTest&test_method=test_basic                               |
| DatalakeDelayedTranslationTest            | test_basic                                 | {"catalog_type": "rest_jdbc", "cloud_storage_type": 1, "query_engine": "spark"}                             | integration | https://buildkite.com/redpanda/redpanda/builds/75633#019a537c-398e-44b5-84a7-b60f2d295c2d | FLAKY         | 2/21     | upstream reliability is '100.0'. current run reliability is '9.523809523809524'. drift is 90.47619 and the allowed drift is set to 50. The test should FAIL             | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=DatalakeDelayedTranslationTest&test_method=test_basic                               |
| DatalakeDelayedTranslationTest            | test_basic                                 | {"catalog_type": "rest_jdbc", "cloud_storage_type": 1, "query_engine": "spark"}                             | integration | https://buildkite.com/redpanda/redpanda/builds/75633#019a5380-ea77-4f96-8901-87be56c64e3b | FAIL          | 0/21     | The test has failed across all retries                                                                                                                                  | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=DatalakeDelayedTranslationTest&test_method=test_basic                               |
| DatalakeDelayedTranslationTest            | test_basic                                 | {"catalog_type": "nessie", "cloud_storage_type": 1, "query_engine": "trino"}                                | integration | https://buildkite.com/redpanda/redpanda/builds/75633#019a537c-3984-4dfc-8417-c128e7e1fd05 | FLAKY         | 1/21     | upstream reliability is '100.0'. current run reliability is '4.761904761904762'. drift is 95.2381 and the allowed drift is set to 50. The test should FAIL              | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=DatalakeDelayedTranslationTest&test_method=test_basic                               |
| DatalakeDelayedTranslationTest            | test_basic                                 | {"catalog_type": "nessie", "cloud_storage_type": 1, "query_engine": "trino"}                                | integration | https://buildkite.com/redpanda/redpanda/builds/75633#019a5380-ea69-46a9-a79f-b47e8e13ff8d | FLAKY         | 7/21     | upstream reliability is '100.0'. current run reliability is '33.33333333333333'. drift is 66.66667 and the allowed drift is set to 50. The test should FAIL             | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=DatalakeDelayedTranslationTest&test_method=test_basic                               |
| DatalakeDelayedTranslationTest            | test_basic                                 | {"catalog_type": "rest_hadoop", "cloud_storage_type": 1, "query_engine": "trino"}                           | integration | https://buildkite.com/redpanda/redpanda/builds/75633#019a537c-3986-4c78-bd52-008b47d8f269 | FLAKY         | 2/21     | upstream reliability is '100.0'. current run reliability is '9.523809523809524'. drift is 90.47619 and the allowed drift is set to 50. The test should FAIL             | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=DatalakeDelayedTranslationTest&test_method=test_basic                               |
| DatalakeDelayedTranslationTest            | test_basic                                 | {"catalog_type": "rest_hadoop", "cloud_storage_type": 1, "query_engine": "trino"}                           | integration | https://buildkite.com/redpanda/redpanda/builds/75633#019a5380-ea6a-4ce3-aa47-2f924c57f591 | FLAKY         | 2/21     | upstream reliability is '100.0'. current run reliability is '9.523809523809524'. drift is 90.47619 and the allowed drift is set to 50. The test should FAIL             | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=DatalakeDelayedTranslationTest&test_method=test_basic                               |
| DatalakeDelayedTranslationTest            | test_basic                                 | {"catalog_type": "rest_jdbc", "cloud_storage_type": 1, "query_engine": "trino"}                             | integration | https://buildkite.com/redpanda/redpanda/builds/75633#019a537c-3987-441c-9366-014909806170 | FAIL          | 0/21     | The test has failed across all retries                                                                                                                                  | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=DatalakeDelayedTranslationTest&test_method=test_basic                               |
| DatalakeDelayedTranslationTest            | test_basic                                 | {"catalog_type": "rest_jdbc", "cloud_storage_type": 1, "query_engine": "trino"}                             | integration | https://buildkite.com/redpanda/redpanda/builds/75633#019a5380-ea6c-4656-a5d8-83804b1b674d | FLAKY         | 3/21     | upstream reliability is '100.0'. current run reliability is '14.285714285714285'. drift is 85.71429 and the allowed drift is set to 50. The test should FAIL            | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=DatalakeDelayedTranslationTest&test_method=test_basic                               |
| DeleteRecordsTest                         | test_delete_records_concurrent_truncations | {"cloud_storage_enabled": false, "truncate_point": "at_high_watermark"}                                     | integration | https://buildkite.com/redpanda/redpanda/builds/75633#019a537c-3988-430c-baa9-e25d72f9dc19 | FLAKY         | 13/21    | upstream reliability is '100.0'. current run reliability is '61.904761904761905'. drift is 38.09524 and the allowed drift is set to 50. The test should PASS            | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=DeleteRecordsTest&test_method=test_delete_records_concurrent_truncations            |
| AWSRoleFetchTests                         | test_write                                 | null                                                                                                        | integration | https://buildkite.com/redpanda/redpanda/builds/75633#019a537c-398c-4fc0-a8df-a573f6d6447e | FLAKY         | 4/21     |                                                                                                                                                                         | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=AWSRoleFetchTests&test_method=test_write                                            |
| STSRoleFetchTests                         | test_write                                 | null                                                                                                        | integration | https://buildkite.com/redpanda/redpanda/builds/75633#019a537c-398d-4381-b5a7-b38dbbeb198a | FLAKY         | 8/21     | upstream reliability is '100.0'. current run reliability is '38.095238095238095'. drift is 61.90476 and the allowed drift is set to 50. The test should FAIL            | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=STSRoleFetchTests&test_method=test_write                                            |
| EndToEndShadowIndexingTest                | test_write                                 | {"cloud_storage_type": 2}                                                                                   | integration | https://buildkite.com/redpanda/redpanda/builds/75633#019a537c-398e-44b5-84a7-b60f2d295c2d | FLAKY         | 4/21     | upstream reliability is '100.0'. current run reliability is '19.047619047619047'. drift is 80.95238 and the allowed drift is set to 50. The test should FAIL            | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=EndToEndShadowIndexingTest&test_method=test_write                                   |
| EndToEndShadowIndexingTestWithDisruptions | test_write_with_node_failures              | {"cloud_storage_type": 2}                                                                                   | integration | https://buildkite.com/redpanda/redpanda/builds/75633#019a537c-398a-4573-8efc-1c6de4ed0f72 | FLAKY         | 11/21    | upstream reliability is '100.0'. current run reliability is '52.38095238095239'. drift is 47.61905 and the allowed drift is set to 50. The test should PASS             | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=EndToEndShadowIndexingTestWithDisruptions&test_method=test_write_with_node_failures |
| FollowerFetchingTest                      | test_with_leadership_transfers             | {"fetch_from": "fetch-from-tiered-storage"}                                                                 | integration | https://buildkite.com/redpanda/redpanda/builds/75633#019a537c-398c-4fc0-a8df-a573f6d6447e | FAIL          | 0/2      |                                                                                                                                                                         | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=FollowerFetchingTest&test_method=test_with_leadership_transfers                     |
| NodeWiseRecoveryTest                      | test_node_wise_recovery                    | {"dead_node_count": 1}                                                                                      | integration | https://buildkite.com/redpanda/redpanda/builds/75633#019a537c-398b-4ea4-bd34-37f2f5e10458 | FAIL          | 0/21     | The test has failed across all retries                                                                                                                                  | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=NodeWiseRecoveryTest&test_method=test_node_wise_recovery                            |
| NodeWiseRecoveryTest                      | test_node_wise_recovery                    | {"dead_node_count": 2}                                                                                      | integration | https://buildkite.com/redpanda/redpanda/builds/75633#019a537c-398c-4fc0-a8df-a573f6d6447e | FAIL          | 0/20     |                                                                                                                                                                         | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=NodeWiseRecoveryTest&test_method=test_node_wise_recovery                            |
| RedpandaNodeOperationsSmokeTest           | test_node_ops_smoke_test                   | {"cloud_storage_type": 1, "mixed_versions": false}                                                          | integration | https://buildkite.com/redpanda/redpanda/builds/75633#019a5380-ea6d-44e7-86ed-d0a11263ff14 | FAIL          | 0/21     | The test has failed across all retries                                                                                                                                  | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=RedpandaNodeOperationsSmokeTest&test_method=test_node_ops_smoke_test                |
| RedpandaNodeOperationsSmokeTest           | test_node_ops_smoke_test                   | {"cloud_storage_type": 1, "mixed_versions": true}                                                           | integration | https://buildkite.com/redpanda/redpanda/builds/75633#019a537c-3989-46c1-971f-b44aeab39816 | FLAKY         | 7/21     | upstream reliability is '95.17045454545455'. current run reliability is '33.33333333333333'. drift is 61.83712 and the allowed drift is set to 50. The test should FAIL | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=RedpandaNodeOperationsSmokeTest&test_method=test_node_ops_smoke_test                |
| ShadowIndexingCompactedTopicTest          | test_upload                                | {"cloud_storage_type": 2}                                                                                   | integration | https://buildkite.com/redpanda/redpanda/builds/75633#019a537c-3989-46c1-971f-b44aeab39816 | FAIL          | 0/21     | The test has failed across all retries                                                                                                                                  | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=ShadowIndexingCompactedTopicTest&test_method=test_upload                            |
| ShadowIndexingCompactedTopicTest          | test_upload                                | {"cloud_storage_type": 1}                                                                                   | integration | https://buildkite.com/redpanda/redpanda/builds/75633#019a537c-398a-4573-8efc-1c6de4ed0f72 | FAIL          | 0/21     | The test has failed across all retries                                                                                                                                  | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=ShadowIndexingCompactedTopicTest&test_method=test_upload                            |
| ShadowLinkingRandomOpsTest                | test_node_operations                       | {"failures": true}                                                                                          | integration | https://buildkite.com/redpanda/redpanda/builds/75633#019a5380-ea6f-4140-a6d1-97514827e386 | FLAKY         | 14/21    | upstream reliability is '100.0'. current run reliability is '66.66666666666666'. drift is 33.33333 and the allowed drift is set to 50. The test should PASS             | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=ShadowLinkingRandomOpsTest&test_method=test_node_operations                         |
| TestTieredStoragePause                    | test_safe_pause_resume                     | {"allow_gaps_cluster_level": true, "allow_gaps_topic_level": true}                                          | integration | https://buildkite.com/redpanda/redpanda/builds/75633#019a5380-ea6d-44e7-86ed-d0a11263ff14 | FLAKY         | 16/21    | upstream reliability is '100.0'. current run reliability is '76.19047619047619'. drift is 23.80952 and the allowed drift is set to 50. The test should PASS             | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=TestTieredStoragePause&test_method=test_safe_pause_resume                           |
| WriteCachingFailureInjectionE2ETest       | test_crash_all                             | {"use_transactions": false}                                                                                 | integration | https://buildkite.com/redpanda/redpanda/builds/75633#019a5380-ea6d-44e7-86ed-d0a11263ff14 | FLAKY         | 17/21    | upstream reliability is '89.08098271155596'. current run reliability is '80.95238095238095'. drift is 8.1286 and the allowed drift is set to 50. The test should PASS   | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=WriteCachingFailureInjectionE2ETest&test_method=test_crash_all                      |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/75732">build#75732</a></summary>


| test_class                           | test_method                   | test_arguments                                                               | test_kind   | job_url                                                                                   | test_status   | passed   | reason                                                                                                                                                                  | test_history                                                                                                                                                                  |
|:-------------------------------------|:------------------------------|:-----------------------------------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|:------------------------------------------------------------------------------------------------------------------------------------------------------------------------|:------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| ShadowLinkConsumeGroupsMirroringTest | test_continuous_group_sync    | {"source_cluster_spec": {"cluster_type": "redpanda"}, "with_failures": true} | integration | https://buildkite.com/redpanda/redpanda/builds/75732#019a5889-4b7c-463c-a6ab-5950714c0b25 | FLAKY         | 20/21    | upstream reliability is '99.11190053285968'. current run reliability is '95.23809523809523'. drift is 3.87381 and the allowed drift is set to 50. The test should PASS  | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=ShadowLinkConsumeGroupsMirroringTest&test_method=test_continuous_group_sync |
| MountUnmountIcebergTest              | test_simple_remount           | {"cloud_storage_type": 1}                                                    | integration | https://buildkite.com/redpanda/redpanda/builds/75732#019a5889-4b7a-490a-a28d-1b7992fd9cf6 | FLAKY         | 15/21    | upstream reliability is '95.33777354900094'. current run reliability is '71.42857142857143'. drift is 23.9092 and the allowed drift is set to 50. The test should PASS  | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=MountUnmountIcebergTest&test_method=test_simple_remount                     |
| LogCompactionTxRemovalTest           | test_tx_control_batch_removal | null                                                                         | integration | https://buildkite.com/redpanda/redpanda/builds/75732#019a587f-c929-4084-989c-d6107c5398b1 | FLAKY         | 15/21    | upstream reliability is '85.87987355110643'. current run reliability is '71.42857142857143'. drift is 14.4513 and the allowed drift is set to 50. The test should PASS  | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=LogCompactionTxRemovalTest&test_method=test_tx_control_batch_removal        |
| WriteCachingFailureInjectionTest     | test_unavoidable_data_loss    | null                                                                         | integration | https://buildkite.com/redpanda/redpanda/builds/75732#019a5889-4b7e-4a79-937c-d4e7b0eeecac | FLAKY         | 20/21    | upstream reliability is '94.70672389127324'. current run reliability is '95.23809523809523'. drift is -0.53137 and the allowed drift is set to 50. The test should PASS | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=WriteCachingFailureInjectionTest&test_method=test_unavoidable_data_loss     |
| TxUpgradeCompactionTest              | upgrade_with_compaction_test  | null                                                                         | integration | https://buildkite.com/redpanda/redpanda/builds/75732#019a5889-4b78-4e4d-bbfd-cf399583670f | FLAKY         | 18/21    | upstream reliability is '98.75776397515527'. current run reliability is '85.71428571428571'. drift is 13.04348 and the allowed drift is set to 50. The test should PASS | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=TxUpgradeCompactionTest&test_method=upgrade_with_compaction_test            |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/75814">build#75814</a></summary>


| test_class                 | test_method                   | test_arguments   | test_kind   | job_url                                                                                   | test_status   | passed   | reason                                                                                                                                                                  | test_history                                                                                                                                                           |
|:---------------------------|:------------------------------|:-----------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|:------------------------------------------------------------------------------------------------------------------------------------------------------------------------|:-----------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| LogCompactionTxRemovalTest | test_tx_control_batch_removal | null             | integration | https://buildkite.com/redpanda/redpanda/builds/75814#019a5d59-08e3-4216-915e-58548e574743 | FLAKY         | 36/40    | upstream reliability is '99.51159951159951'. current run reliability is '83.33333333333334'. drift is 16.17827 and the allowed drift is set to 50. The test should PASS | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=LogCompactionTxRemovalTest&test_method=test_tx_control_batch_removal |
| LogCompactionTxRemovalTest | test_tx_control_batch_removal | null             | integration | https://buildkite.com/redpanda/redpanda/builds/75814#019a5d59-08e4-4970-99d9-70a73f0827ae | FLAKY         | 39/40    | upstream reliability is '99.51159951159951'. current run reliability is '95.23809523809523'. drift is 4.2735 and the allowed drift is set to 50. The test should PASS   | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=LogCompactionTxRemovalTest&test_method=test_tx_control_batch_removal |
| LogCompactionTxRemovalTest | test_tx_control_batch_removal | null             | integration | https://buildkite.com/redpanda/redpanda/builds/75814#019a5d59-08e5-4863-938c-5e48676ceeba | FLAKY         | 34/40    | upstream reliability is '99.51159951159951'. current run reliability is '75.0'. drift is 24.5116 and the allowed drift is set to 50. The test should PASS               | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=LogCompactionTxRemovalTest&test_method=test_tx_control_batch_removal |
| LogCompactionTxRemovalTest | test_tx_control_batch_removal | null             | integration | https://buildkite.com/redpanda/redpanda/builds/75814#019a5d59-08e6-463e-aa79-bcc7058c24bd | FLAKY         | 36/40    | upstream reliability is '99.51159951159951'. current run reliability is '81.81818181818183'. drift is 17.69342 and the allowed drift is set to 50. The test should PASS | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=LogCompactionTxRemovalTest&test_method=test_tx_control_batch_removal |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/76098">build#76098</a></summary>


| test_class                        | test_method                                | test_arguments                                 | test_kind   | job_url                                                                                   | test_status   | passed   | reason                                                                                                                                                                 | test_history                                                                                                                                                                               |
|:----------------------------------|:-------------------------------------------|:-----------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|:-----------------------------------------------------------------------------------------------------------------------------------------------------------------------|:-------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| ShadowLinkingReplicationTests     | test_replication_with_failures             | null                                           | integration | https://buildkite.com/redpanda/redpanda/builds/76098#019a7779-5b17-4a60-afbf-e5c8fcc0ad31 | FLAKY         | 19/21    | upstream reliability is '96.31490787269682'. current run reliability is '90.47619047619048'. drift is 5.83872 and the allowed drift is set to 50. The test should PASS | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=ShadowLinkingReplicationTests&test_method=test_replication_with_failures                 |
| LogCompactionTxRemovalUpgradeTest | test_tx_control_batch_removal_with_upgrade | {"test_case_name": "Mixed aborts and commits"} | integration | https://buildkite.com/redpanda/redpanda/builds/76098#019a7774-1c73-4975-b3f2-e1fdfae76c4a | FLAKY         | 20/21    | upstream reliability is '100.0'. current run reliability is '95.23809523809523'. drift is 4.7619 and the allowed drift is set to 50. The test should PASS              | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=LogCompactionTxRemovalUpgradeTest&test_method=test_tx_control_batch_removal_with_upgrade |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/76132">build#76132</a></summary>


| test_class                         | test_method                   | test_arguments                                                                    | test_kind   | job_url                                                                                   | test_status   | passed   | reason                                                                                                                                                                  | test_history                                                                                                                                                                   |
|:-----------------------------------|:------------------------------|:----------------------------------------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|:------------------------------------------------------------------------------------------------------------------------------------------------------------------------|:-------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| ShadowLinkingReplicationTests      | test_replication_basic        | {"shuffle_leadership": true, "source_cluster_spec": {"cluster_type": "redpanda"}} | integration | https://buildkite.com/redpanda/redpanda/builds/76132#019a78f6-5ffe-427d-989e-73806037ffbd | FLAKY         | 16/21    | upstream reliability is '88.78748370273793'. current run reliability is '76.19047619047619'. drift is 12.59701 and the allowed drift is set to 50. The test should PASS | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=ShadowLinkingReplicationTests&test_method=test_replication_basic             |
| ControllerLogLimitMirrorMakerTests | test_mirror_maker_with_limits | null                                                                              | integration | https://buildkite.com/redpanda/redpanda/builds/76132#019a78f4-f6c5-439c-becf-7ce92f97c3fc | FLAKY         | 20/21    | upstream reliability is '98.95833333333334'. current run reliability is '95.23809523809523'. drift is 3.72024 and the allowed drift is set to 50. The test should PASS  | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=ControllerLogLimitMirrorMakerTests&test_method=test_mirror_maker_with_limits |
</details>
<!--/body-->

mmaslankaprv committed:
c/rm_stm: do not update LSO when begin/end tx is in progress

This commit introduces a new read/write lock that protects the LSO updates.
The lock is held for write when there is any operation that when applied
may influence calculation of LSO i.e. begin_tx and end transaction.

The lock is being acquired for read when `rm_stm` is being asked about
the LSO. If the lock is not acquired - a transaction operation is in
progress we fallback to previous LSO, otherwise the calculation is based
on the state that is guaranteed to be up to date.

Signed-off-by: Michał Maślanka <michal@redpanda.com>

mmaslankaprv committed:
Revert "`tree-wide`: disable compaction of transactional control batches"

This reverts commit 94c2b012ad4b022369f00437c452a5ab2f922e10.

mmaslankaprv commented:
/ci-repeat 1

mmaslankaprv commented:
/ci-repeat 5
skip-redpanda-build
dt-repeat=20
tests/rptest/tests/log_compaction_test.py::LogCompactionTxRemovalTest

WillemKauf reviewed (dismissed):
Awesome :partying_face:

bharathv commented on src/v/cluster/rm_stm.h:
```diff
@@ -435,6 +435,7 @@ class rm_stm final : public raft::persisted_stm<> {
     model::producer_id _highest_producer_id;
     // for monotonicity of computed LSO.
     model::offset _last_known_lso{-1};
+    ss::rwlock _lso_lock;
```
nit: could use a comment.

bharathv commented on src/v/cluster/rm_stm.cc:
```diff
@@ -660,6 +662,7 @@ ss::future<tx::errc> rm_stm::abort_tx(
   model::tx_seq tx_seq,
   model::timeout_clock::duration timeout) {
     auto state_lock_holder = co_await _state_lock.hold_read_lock();
+    auto lso_lock_holder = co_await _lso_lock.hold_write_lock();
```
do we need this lock in the commit/abort path? (LSO is clamped by fence always?)

bharathv commented on src/v/cluster/rm_stm.cc:
```diff
@@ -1305,6 +1308,16 @@ model::offset rm_stm::last_stable_offset() {
+          "lso update in progress, last_known_lso: {}, last_applied: {}",
+          _last_known_lso,
+          last_applied);
+        return _last_known_lso;
```
One minor optimization is to advance lso right after obtaining _lso_lock.write_lock(). Doing so ensures that lso reflects the last_visible_index at that time, instead of depending on the previous lso() call, which might be outdated.

mmaslankaprv replied:
I am not sure i understand that part

bharathv replied:
Consider this situation: when last_stable_offset() is called at time `t`, the `_last_known_lso` is set to 100.
Now, suppose several transactions have occurred, and the actual LSO is now at 900. When the new begin (at offset 901) and last_stable_offset() are executed concurrently at say `t + 1hr`, a race condition occurs and runs into this check and the function conservatively returns 100, since that’s the last recorded known LSO.
Ideally, the function could return 900 instead. One possible solution would be to recompute the _last_known_offset after acquiring the _lso_lock in begin, but before performing replication, I think.

bharathv reviewed (commented):
nice idea, couple of questions.

joe-redpanda commented on src/v/cluster/rm_stm.cc:
```diff
@@ -1852,14 +1852,15 @@ model::offset rm_stm::to_log_offset(kafka::offset k_offset) const {
 
 ss::future<raft::local_snapshot_applied>
 rm_stm::apply_local_snapshot(raft::stm_snapshot_header hdr, iobuf&& tx_ss_buf) {
+    auto data_buf = std::move(tx_ss_buf);
```
This doesn't work if preemption check decides to yield before apply_local_snapshot.

We should probably fix this everywhere sooner rather than later.

alternative to definitely fix rm_stm for now can we:

rm_stm::apply_local_snapshot(raft::stm_snapshot_header hdr, iobuf&& tx_ss_buf) {
    return do_apply_local_snapshot(std::move(hdr), std::move(tx_ss_buf));
}

rm_stm::do_apply_local_snapshot(raft::stm_snapshot_header hdr, iobuf tx_ss_buf) {
    .... // the rest of the implementation
}

bashtanov replied:
Where exactly preemption check can yield? When the function is called like `... = co_await apply_local_snapshot(...)` the initial synchronous stretch of the function before the first scheduling point is executed immediately. A function call, whether the functions returns a future or not, and whether or not it is a coroutine, does not form a scheduling point.

mmaslankaprv replied:
i think this is fine

joe-redpanda replied:
seastar coroutines aren't always eager, I'll link the thread

I won't block on this comment because this clearly does not make the issue worse

mmaslankaprv commented:
/ci-repeat 1

bashtanov commented on src/v/cluster/tests/rm_stm_tests.cc:
```diff
@@ -981,3 +981,89 @@ FIXTURE_TEST(test_concurrent_producer_evictions, rm_stm_test_fixture) {
+        random_sleep().get();
+        auto result = replicate_all(stm, make_batches(pid, 0, 5, true)).get();
+        BOOST_REQUIRE(result.has_value());
+        // commit
```
does this comment belong to the last 2 statements in the lambda rather than here?

bashtanov commented on src/v/cluster/tests/rm_stm_tests.cc:
```diff
@@ -981,3 +981,89 @@ FIXTURE_TEST(test_concurrent_producer_evictions, rm_stm_test_fixture) {
+          });
+    };
+
+    std::optional<model::offset> earliest_open_tx;
```
nit: this variable tracks no more than one open transaction, so I don't like the word `earliest` in it's name. Maybe rename to `open_tx_start_offset` or something like this?

bashtanov commented on src/v/cluster/tests/rm_stm_tests.cc:
```diff
@@ -981,3 +981,89 @@ FIXTURE_TEST(test_concurrent_producer_evictions, rm_stm_test_fixture) {
+      [&as] { return as.abort_requested(); },
+      [&] {
+          auto current_lso = stm.last_stable_offset();
+          auto lso_bound = !earliest_open_tx || current_lso == earliest_open_tx;
```
nit: `bool` is no longer to type than `auto`, but would  improve readability, as "bound" may be read a noun as well as a participle

bashtanov commented on src/v/cluster/rm_stm.cc:
```diff
@@ -1331,6 +1331,37 @@ model::offset rm_stm::last_stable_offset() {
+        //
+        // The issue here if a begin_tx is replicated but not yet applied
+        // in the stm and a caller queries LSO, we return LSO beyond the
+        // begin_tx which is incorrect. The problem is _active_tx_producers
```
nit: Something is wrong with this sentence (or with my ability to read English). `The issue here IS THAT ...`?

bashtanov commented on src/v/cluster/rm_stm.cc:
```diff
@@ -1331,6 +1331,37 @@ model::offset rm_stm::last_stable_offset() {
+        // The impact of this overestimation is that compaction may compact
+        // away open transaction begin marker as it relies on LSO. The
+        // chances are rare but not impossible :(. if at that point the replica
+        // restarts and there are no furher updates in the transaction, the
```
typo: further

bashtanov reviewed (dismissed):
LGTM, nits only

WillemKauf reviewed (approved):
Looks good. See thread here: https://redpandadata.slack.com/archives/C07FJGU5AKV/p1763062534096009 on discussion for release of the feature (i.e are we backporting to v25.3.x for release in a future minor?)
